From 2ccf7879b9e4feb093d2acc6859edce53b7cdfd2 Mon Sep 17 00:00:00 2001
From: Allen Xu <b45815@freescale.com>
Date: Thu, 4 Dec 2014 01:06:47 +0800
Subject: [PATCH] MLK-9949: mtd: qspi: Handle QSPI probe error path correctly

If QSPI probe failed in some cases, such as board rework, the error
patch was not handled correctly.

This issue may cause kernle dump in fec driver, since the
pm_qos_remove_request() in QSPI driver was not invoked when probe
failed.

Signed-off-by: Allen Xu <b45815@freescale.com>
(cherry picked from commit 350d532e0266a0a6918cbc6a17952ef64aef2521)
---
 drivers/mtd/spi-nor/fsl-quadspi.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/mtd/spi-nor/fsl-quadspi.c b/drivers/mtd/spi-nor/fsl-quadspi.c
index c5aa85a..809cf1a 100644
--- a/drivers/mtd/spi-nor/fsl-quadspi.c
+++ b/drivers/mtd/spi-nor/fsl-quadspi.c
@@ -1018,16 +1018,18 @@ static int fsl_qspi_probe(struct platform_device *pdev)
 
 		ret = of_modalias_node(np, modalias, sizeof(modalias));
 		if (ret < 0)
-			goto map_failed;
+			goto irq_failed;
 
 		id = spi_nor_match_id(modalias);
-		if (!id)
-			goto map_failed;
+		if (!id) {
+			ret = -EINVAL;
+			goto irq_failed;
+		}
 
 		ret = of_property_read_u32(np, "spi-max-frequency",
 				&q->clk_rate);
 		if (ret < 0)
-			goto map_failed;
+			goto irq_failed;
 
 		/* Can we enable the DDR Quad Read? */
 		ret = of_property_read_u32(np, "spi-nor,ddr-quad-read-dummy",
@@ -1040,12 +1042,12 @@ static int fsl_qspi_probe(struct platform_device *pdev)
 
 		ret = spi_nor_scan(nor, id, mode);
 		if (ret)
-			goto map_failed;
+			goto irq_failed;
 
 		ppdata.of_node = np;
 		ret = mtd_device_parse_register(mtd, NULL, &ppdata, NULL, 0);
 		if (ret)
-			goto map_failed;
+			goto irq_failed;
 
 		/* Set the correct NOR size now. */
 		if (q->nor_size == 0) {
-- 
1.8.0

